pkg/controller/oneagent/oneagent_update.go

```diff
@@ -0,0 +1,211 @@
+)
+
+func (r *ReconcileOneAgent) reconcileVersion(logger logr.Logger, instance dynatracev1alpha1.BaseOneAgentDaemonSet, dtc dtclient.Client) (bool, error) {
+	if instance.GetOneAgentSpec().UseImmutableImage {
+		return r.reconcileVersionImmutableImage(instance)
+	} else {
+		return r.reconcileVersionInstaller(logger, instance, dtc)
+	}
+}
+
+func (r *ReconcileOneAgent) reconcileVersionInstaller(logger logr.Logger, instance dynatracev1alpha1.BaseOneAgentDaemonSet, dtc dtclient.Client) (bool, error) {
+	updateCR := false
+
+	desired, err := dtc.GetLatestAgentVersion(dtclient.OsUnix, dtclient.InstallerTypeDefault)
+	if err != nil {
+		return false, fmt.Errorf("failed to get desired version: %w", err)
+	} else if desired != "" && instance.GetOneAgentStatus().Version != desired {
+		logger.Info("new version available", "actual", instance.GetOneAgentStatus().Version, "desired", desired)
+		instance.GetOneAgentStatus().Version = desired
+		updateCR = true
+	}
+
+	podList, err := r.findPods(instance)
+	if err != nil {
+		logger.Error(err, "failed to list pods", "podList", podList)
+		return updateCR, err
+	}
+
+	podsToDelete, instances, err := findOutdatedPodsInstaller(podList, dtc, instance)
+	if err != nil {
+		return updateCR, err
+	}
+
+	// Workaround: 'instances' can be null, making DeepEqual() return false when comparing against an empty map instance.
+	// So, compare as long there is data.
+	if (len(instances) > 0 || len(instance.GetOneAgentStatus().Instances) > 0) && !reflect.DeepEqual(instances, instance.GetOneAgentStatus().Instances) {
+		logger.Info("oneagent pod instances changed", "status", instance.GetOneAgentStatus())
+		updateCR = true
+		instance.GetOneAgentStatus().Instances = instances
+	}
+
+	var waitSecs uint16 = 300
+	if instance.GetOneAgentSpec().WaitReadySeconds != nil {
+		waitSecs = *instance.GetOneAgentSpec().WaitReadySeconds
+	}
+
+	if len(podsToDelete) > 0 {
+		if instance.GetOneAgentStatus().SetPhase(dynatracev1alpha1.Deploying) {
+			err := r.updateCR(instance)
+			if err != nil {
+				logger.Error(err, fmt.Sprintf("failed to set phase to %s", dynatracev1alpha1.Deploying))
+			}
+		}
+	}
+
+	// restart daemonset
+	err = r.deletePods(logger, podsToDelete, buildLabels(instance.GetName()), waitSecs)
+	if err != nil {
+		logger.Error(err, "failed to update version")
+		return updateCR, err
+	}
+
+	return updateCR, nil
+}
+
+func (r *ReconcileOneAgent) reconcileVersionImmutableImage(instance dynatracev1alpha1.BaseOneAgentDaemonSet) (bool, error) {
+	updateCR := false
+	var waitSecs uint16 = 300
+	if instance.GetOneAgentSpec().WaitReadySeconds != nil {
+		waitSecs = *instance.GetOneAgentSpec().WaitReadySeconds
+	}
+
+	if !instance.GetOneAgentSpec().DisableAgentUpdate {
+		r.logger.Info("checking for outdated pods")
+		// Check if pods have latest agent version
+		outdatedPods, err := r.findOutdatedPodsImmutableImage(r.logger, instance, isLatest)
+		if len(outdatedPods) > 0 {
+			updateCR = true
+		}
+		if err != nil {
```

Comment:
Minor: I'd suggest checking for errors before the result.

Reply:
Moved